server/backend/feedback.go

```diff
@@ -0,0 +1,142 @@
+package backend
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path"
+
+	pb "github.com/TUM-Dev/Campus-Backend/server/api/tumdev"
+	"github.com/TUM-Dev/Campus-Backend/server/backend/cron"
+	"github.com/TUM-Dev/Campus-Backend/server/model"
+	"github.com/gabriel-vasile/mimetype"
+	"github.com/gofrs/uuid/v5"
+	"github.com/guregu/null"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gorm.io/gorm"
+)
+
+// NewFeedback accepts a stream of feedback messages from the client and stores them in the database/file system.
+func (s *CampusServer) NewFeedback(stream pb.Campus_NewFeedbackServer) error {
+	return s.db.WithContext(stream.Context()).Transaction(func(tx *gorm.DB) error {
+		// receive metadata
+		imageCount := int32(0)
+		id, err := uuid.NewGen().NewV7()
+		if err != nil {
+			log.WithError(err).Error("Error generating uuid")
+			return status.Error(codes.Internal, "Error starting feedback submission")
+		}
+		feedback := &model.Feedback{EmailId: null.StringFrom(id.String())}
+
+		// download images
+		for {
+			req, err := stream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				log.WithError(err).Error("Error receiving feedback")
+				deleteUploaded(tx, feedback.EmailId.String)
+				return status.Error(codes.Internal, "Error receiving feedback")
+			}
+			mergeFeedback(feedback, req)
+
+			if len(req.Attachment) > 0 {
+				imageCount = handleImageUpload(tx, &req.Attachment, imageCount, feedback.EmailId.String)
+			}
+		}
+		feedback.ImageCount = imageCount
+		if err := tx.Create(feedback).Error; err != nil {
+			log.WithError(err).Error("Error creating feedback")
+			return status.Error(codes.Internal, "Error creating feedback")
+		}
+		if err := stream.SendAndClose(&pb.NewFeedbackReply{}); err != nil {
+			log.WithError(err).Error("Error sending feedbackreply")
+			return status.Error(codes.Internal, "Error sending feedbackreply")
+		}
+		return nil
+	})
+}
+
+func deleteUploaded(tx *gorm.DB, id string) {
+	// delete uploaded images
+	dbPath := fmt.Sprintf("feedback/%s", id)
```

Comment:
How does this work with multiple images? (Like its fine if we only want to make it work with a single image)

Reply:
Deleting a directory deletes all contained images

Reply:
oh, its a folder :facepalm: